python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -249,23 +265,40 @@ def read_messages(
         """
         with self.get_params() as params:
             is_task_complete = Event()
-            thread = None
+            messages_thread = None
+            stdout_thread = None
+            stderr_thread = None
             try:
-                thread = Thread(
-                    target=self._reader_thread,
-                    args=(
-                        handler,
-                        params,
-                        is_task_complete,
-                    ),
-                    daemon=True,
+                messages_thread = Thread(
+                    target=self._messages_thread, args=(handler, params, is_task_complete)
                 )
-                thread.start()
+                messages_thread.start()
+                if self.stdout_reader:
+                    stdout_thread = self.stdout_reader.start_thread(params, is_task_complete)
+                if self.stderr_reader:
+                    stderr_thread = self.stderr_reader.start_thread(params, is_task_complete)
                 yield params
             finally:
+                self.wait_for_stdio_logs(params)
                 is_task_complete.set()
-                if thread:
-                    thread.join()
+                if messages_thread:
+                    messages_thread.join()
+                if stdout_thread:
+                    stdout_thread.join()
+                if stderr_thread:
+                    stderr_thread.join()
```

Comment:
Hmm, seems like the `PipesBlobStoreStdioReader` should encapsulate the thread completely?

Reply:
updated accordingly